From 7390f2154704c015250c6309e8c219ff622a327c Mon Sep 17 00:00:00 2001
From: Potnuri Bharat Teja <bharat@chelsio.com>
Date: Wed, 13 Mar 2019 15:19:59 +0530
Subject: [PATCH 127/127] iw_cxgb4: always disconnect when transitioning to
 TERMINATE

commit d2c33370ae73105c7c7df8f7048d20653991b4cb
Author: Potnuri Bharat Teja <bharat@chelsio.com>
Date:   Tue Apr 2 14:46:11 2019 +0530

    RDMA/iw_cxgb4: Always disconnect when QP is transitioning to TERMINATE state

    On receiving a TERM from tje peer, Host moves the QP to TERMINATE state
    and then moves the adapter out of RDMA mode. After issuing a TERM, peer
    issues a CLOSE and at this point of time if the connectivity between peer
    and host is lost for a significant amount of time, the QP remains in
    TERMINATE state.

    Therefore c4iw_modify_qp() needs to initiate a close on entering terminate
    state.

    Signed-off-by: Potnuri Bharat Teja <bharat@chelsio.com>
    Signed-off-by: Jason Gunthorpe <jgg@mellanox.com>

Signed-off-by: Potnuri Bharat Teja <bharat@chelsio.com>
---
 drivers/infiniband/hw/cxgb4/qp.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/infiniband/hw/cxgb4/qp.c b/drivers/infiniband/hw/cxgb4/qp.c
index 6119a3bcdbe7..42d2e26c802d 100644
--- a/drivers/infiniband/hw/cxgb4/qp.c
+++ b/drivers/infiniband/hw/cxgb4/qp.c
@@ -1670,10 +1670,10 @@ int c4iw_modify_qp(struct c4iw_dev *rhp, struct c4iw_qp *qhp,
 			qhp->attr.layer_etype = attrs->layer_etype;
 			qhp->attr.ecode = attrs->ecode;
 			ep = qhp->ep;
+			c4iw_get_ep(&ep->com);
+			disconnect = 1;
 			if (!internal) {
-				c4iw_get_ep(&qhp->ep->com);
 				terminate = 1;
-				disconnect = 1;
 			} else {
 				terminate = qhp->attr.send_term;
 				ret = rdma_fini(rhp, qhp, ep);
-- 
2.18.0.232.gb7bd9486b055

